File: frontend/lib/data/services/supplier_service.dart
import 'package:dio/dio.dart';
import '../models/supplier.dart';
import 'api_service.dart';

/// Service pour gérer les fournisseurs
class SupplierService {
  final ApiService _apiService;

  SupplierService(this._apiService);

  /// Récupérer tous les fournisseurs
  Future<List<Supplier>> getAllSuppliers({String? status, String? search}) async {
    try {
      final queryParams = <String, dynamic>{};
      if (status != null) queryParams['status'] = status;
      if (search != null) queryParams['search'] = search;

      final response = await _apiService.dio.get(
        '/suppliers',
        queryParameters: queryParams,
      );

      if (response.data['success'] == true) {
        final List<dynamic> data = response.data['data'];
        return data.map((json) => Supplier.fromJson(json)).toList();
      }

      throw Exception('Erreur lors de la récupération des fournisseurs');
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }

  /// Récupérer un fournisseur par ID
  Future<Supplier> getSupplierById(int id) async {
    try {
      final response = await _apiService.dio.get('/suppliers/$id');

      if (response.data['success'] == true) {
        return Supplier.fromJson(response.data['data']);
      }

      throw Exception('Fournisseur introuvable');
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }

  /// Créer un nouveau fournisseur (Admin)
  Future<Supplier> createSupplier({
    required String name,
    String? email,
    String? phone,
    String? address,
    String status = 'active',
  }) async {
    try {
      final response = await _apiService.dio.post(
        '/suppliers',
        data: {
          'name': name,
          'email': email,
          'phone': phone,
          'address': address,
          'status': status,
        },
      );

      if (response.data['success'] == true) {
        return Supplier.fromJson(response.data['data']);
      }

      throw Exception('Erreur lors de la création du fournisseur');
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }

  /// Mettre à jour un fournisseur (Admin)
  Future<Supplier> updateSupplier({
    required int id,
    String? name,
    String? email,
    String? phone,
    String? address,
    String? status,
  }) async {
    try {
      final data = <String, dynamic>{};
      if (name != null) data['name'] = name;
      if (email != null) data['email'] = email;
      if (phone != null) data['phone'] = phone;
      if (address != null) data['address'] = address;
      if (status != null) data['status'] = status;

      final response = await _apiService.dio.put(
        '/suppliers/$id',
        data: data,
      );

      if (response.data['success'] == true) {
        return Supplier.fromJson(response.data['data']);
      }

      throw Exception('Erreur lors de la mise à jour du fournisseur');
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }

  /// Supprimer un fournisseur (Admin)
  Future<void> deleteSupplier(int id) async {
    try {
      final response = await _apiService.dio.delete('/suppliers/$id');

      if (response.data['success'] != true) {
        throw Exception('Erreur lors de la suppression du fournisseur');
      }
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }

  /// Changer le statut d'un fournisseur (Admin)
  Future<Supplier> changeStatus(int id, String status) async {
    try {
      final response = await _apiService.dio.patch(
        '/suppliers/$id/status',
        data: {'status': status},
      );

      if (response.data['success'] == true) {
        return Supplier.fromJson(response.data['data']);
      }

      throw Exception('Erreur lors du changement de statut');
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? 'Erreur réseau');
    }
  }
}

File: backend/src/services/supplierService.js
const pool = require('../config/database');
const Supplier = require('../models/Supplier');

/**
 * Service pour gérer les fournisseurs
 */
class SupplierService {
  /**
   * Récupérer tous les fournisseurs
   */
  async getAll(filters = {}) {
    try {
      let query = 'SELECT * FROM suppliers';
      const params = [];
      const conditions = [];

      // Filtrer par statut
      if (filters.status) {
        conditions.push(`status = $${params.length + 1}`);
        params.push(filters.status);
      }

      // Recherche par nom
      if (filters.search) {
        conditions.push(`name ILIKE $${params.length + 1}`);
        params.push(`%${filters.search}%`);
      }

      if (conditions.length > 0) {
        query += ` WHERE ${conditions.join(' AND ')}`;
      }

      query += ' ORDER BY created_at DESC';

      const result = await pool.query(query, params);
      return result.rows.map(row => Supplier.fromDB(row));
    } catch (error) {
      throw new Error(`Erreur lors de la récupération des fournisseurs: ${error.message}`);
    }
  }

  /**
   * Récupérer un fournisseur par ID
   */
  async getById(id) {
    try {
      const result = await pool.query('SELECT * FROM suppliers WHERE id = $1', [id]);

      if (result.rows.length === 0) {
        throw new Error('Fournisseur introuvable');
      }

      return Supplier.fromDB(result.rows[0]);
    } catch (error) {
      throw error;
    }
  }

  /**
   * Créer un nouveau fournisseur
   */
  async create(supplierData) {
    try {
      const { name, email, phone, address, status = 'active' } = supplierData;

      // Vérifier si l'email existe déjà
      if (email) {
        const existingSupplier = await pool.query('SELECT id FROM suppliers WHERE email = $1', [email]);
        if (existingSupplier.rows.length > 0) {
          throw new Error('Un fournisseur avec cet email existe déjà');
        }
      }

      const result = await pool.query(
        `INSERT INTO suppliers (name, email, phone, address, status)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING *`,
        [name, email, phone, address, status]
      );

      return Supplier.fromDB(result.rows[0]);
    } catch (error) {
      throw error;
    }
  }

  /**
   * Mettre à jour un fournisseur
   */
  async update(id, supplierData) {
    try {
      const { name, email, phone, address, status } = supplierData;

      // Vérifier si le fournisseur existe
      await this.getById(id);

      // Vérifier si l'email est déjà utilisé par un autre fournisseur
      if (email) {
        const existingSupplier = await pool.query(
          'SELECT id FROM suppliers WHERE email = $1 AND id != $2',
          [email, id]
        );
        if (existingSupplier.rows.length > 0) {
          throw new Error('Un autre fournisseur utilise déjà cet email');
        }
      }

      const result = await pool.query(
        `UPDATE suppliers
         SET name = COALESCE($1, name),
             email = COALESCE($2, email),
             phone = COALESCE($3, phone),
             address = COALESCE($4, address),
             status = COALESCE($5, status),
             updated_at = CURRENT_TIMESTAMP
         WHERE id = $6
         RETURNING *`,
        [name, email, phone, address, status, id]
      );

      return Supplier.fromDB(result.rows[0]);
    } catch (error) {
      throw error;
    }
  }

  /**
   * Supprimer un fournisseur
   */
  async delete(id) {
    try {
      // Vérifier si le fournisseur existe
      await this.getById(id);

      await pool.query('DELETE FROM suppliers WHERE id = $1', [id]);

      return { message: 'Fournisseur supprimé avec succès' };
    } catch (error) {
      throw error;
    }
  }

  /**
   * Changer le statut d'un fournisseur
   */
  async changeStatus(id, status) {
    try {
      if (!['active', 'inactive'].includes(status)) {
        throw new Error('Statut invalide. Utilisez "active" ou "inactive"');
      }

      const result = await pool.query(
        `UPDATE suppliers
         SET status = $1, updated_at = CURRENT_TIMESTAMP
         WHERE id = $2
         RETURNING *`,
        [status, id]
      );

      if (result.rows.length === 0) {
        throw new Error('Fournisseur introuvable');
      }

      return Supplier.fromDB(result.rows[0]);
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new SupplierService();
